### DISCUSSION TITLE
smoothOut after Boolean

### DISCUSSION THREAD

**Discussion by davidh71** [elalish/manifold#1073] [Q&A]
Is there a way to avoid mesh weirdness like at the intersection of two joined boxes?
example: https://tinyurl.com/3utshuxr

Appreciate your help.

**Comment by elalish** [comment]
Not really - this comes from the fact that our smoothing interpolates, so each input vertex is unmoved. However, a Boolean operation does intersections with flat triangles, which introduces error if you're wanting a smoothly-curved surface, and tends to create oscillations like you see. Generally I'd recommend doing your smoothing and refining to your required tolerance before any Boolean operations. 

I'd like to someday make a modifier on the Boolean to move the intersected verts to the smoothed surfaces to make this work better, but that'll be tricky in the general case.

**Reply by elalish** [reply]
FYI - it has nothing to do with concavity, but the Boolean operation itself. We handle concave regions just fine, but you may want some kind of chamfer to guide the shape.

**Reply by davidh71** [reply]
Thanks - if i do the smoothing before the boolean, that joined edge will be very sharp.  Ideally a smooth after the boolean would clean up that concave.   I tried doing the boolean first without smooth and then applying a smooth and that led to a very unexpected result (for someone who doesn't understand what it's doing). 

https://manifoldcad.org/#New%20Script%202%201%3Ccode%3Econst%20%7Bcube%2C%20sphere%7D%20%3D%20Manifold%3B%0A%0Aconst%20boxA%20%3D%20cube(%5B10%2C%2010%2C%2010%5D%2C%20true)%0A.rotate(%5B0%2C45%2C0%5D)%3B%0A%0Aconst%20boxB%20%3D%20cube(%5B10%2C%2010%2C%2010%5D%2C%20true)%0A.rotate(%5B0%2C45%2C0%5D).translate(%5B10%2C0%2C0%5D)%3B%0A%0Aconst%20result%20%3D%20boxA.add(boxB).smoothOut(120%2C%201).refine(20)%3B

**Reply by elalish** [reply]
Yes, Booleans are intended to make sharp joints. If you want smooth joints, you might consider SDF-style Booleans instead - see https://github.com/elalish/manifold/blob/master/test/sdf_test.cpp#L202

I'm also interesting in creating a chamfer function, which our smoothing could then turn into fillets. Might be interesting to be able to apply that to Boolean intersections. Not sure when that'll happen though.

**Reply by davidh71** [reply]
Understand that's what booleans do.  What i meant was the smoothOut operation.  Here's an example of the same subdivided cubes, booleaned together with a smooth operator in blender. you can see that interior area also smooths.
![image](https://github.com/user-attachments/assets/d5765e3d-fe28-469f-bf9a-4914ff2ea793)

**Reply by elalish** [reply]
Yeah, so perhaps I should have found better terminology than `smooth` - these are very different operations. Blender is doing something that is probably better to call smooth: moving vertices to smooth the shape. My version moves no vertices - it creates tangents for smooth *interpolation*, which is why you can only see the results once you refine the shape. There is no way to achieve what blender is doing if the original vertices stay in place. 

My version is less about art and more about engineering: making a smoothly-interpolated curve between points whose position is important and meaningful, perhaps for joining to another object.
